### PR TITLE
MapData refactor + Bug Fix

### PR DESCRIPTION
**Pull request by DanVanAtta**
-Has bug fix
- Move Mapdata.java to its own package
- refactors a number of items in MapData.java, many more could be done, but are left for another day at this point.

---
This change is [Reviewable](https://reviewable.io/reviews/triplea-game/triplea/901)

### CODE REVIEW

File: src/games/strategy/triplea/ui/MapData.java [hunk 1]
```diff
@@ -204,11 +204,23 @@ private void loadTerritoryNames() {
     }
     for (final String name : m_centers.keySet()) {
       try {
-        final Image img = loadImage("territoryNames/" + name + ".png");
+        String normalizedName = name.replace(' ', '_');
+        String path = "territoryNames/" + normalizedName + ".png";
+        Image img = loadImage(path);
+
+        if(img == null ) {
+          path = "territoryNames/" + normalizedName + ".png";
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/ui/MapData.java]: Should one of these be trying 'name' instead of 'normalizedName'? It appears its doing the same thing twice.

Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/MapData.java]: Yes, I realized that later though when I was refactoring this class. Originally:
#900 was a straight bug fix
#901 (this PR) was a move
#902 is a PR into #901 - refactoring MapData

I closed #900 since the fix was not right there.  #902 still contains the fix: 3c9282159b1ca173f29f4f2605bf20a73ae3fa49

**Comment by RoiEXLab** [comment]
@DanVanAtta No objections on merging...

File: src/games/strategy/triplea/ui/mapdata/MapData.java [hunk 2]
```diff
@@ -193,43 +188,65 @@ public MapData(final ResourceLoader loader) {
...
-      return;
+  private Optional<Image> loadTerritoryNameImage(String imageName) {
+    Image img = null;
+    try {
+      // try first file names that have underscores instead of spaces
+      final String normalizedName = imageName.replace(' ', '_');
+      img = loadImage(constructTerritoryNameImagePath(normalizedName));
+      if (img == null) {
+        img = loadImage(constructTerritoryNameImagePath(imageName));
+      }
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/ui/mapdata/MapData.java]: Shouldn't we try the actual name before the normalized one?

File: src/games/strategy/triplea/ui/mapdata/MapData.java [hunk 3]
```diff
@@ -173,15 +168,15 @@ public MapData(final ResourceLoader loader) {
       m_namePlace = PointFileReaderWriter.readOneToOne(loader.getResourceAsStream(prefix + TERRITORY_NAME_PLACE_FILE));
       m_kamikazePlace = PointFileReaderWriter.readOneToOne(loader.getResourceAsStream(prefix + KAMIKAZE_FILE));
       m_mapProperties = new Properties();
-      loadDecorations();
-      loadTerritoryNames();
+      m_decorations = loadDecorations();
+      m_territoryNameImages = territoryNameImages();
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/ui/mapdata/MapData.java]: Would be nice to remove the m_* in this class.

File: src/games/strategy/triplea/ui/mapdata/MapData.java [hunk 4]
```diff
@@ -193,43 +188,65 @@ public MapData(final ResourceLoader loader) {
...
   public void close() {
     m_resourceLoader.close();
   }
 
-  private void loadTerritoryNames() {
-    m_territoryNameImages = new HashMap<>();
+  private Map<String, Image> territoryNameImages() {
     if (!m_resourceLoader.hasPath("territoryNames/")) {
-      return;
+      return Collections.EMPTY_MAP;
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/ui/mapdata/MapData.java]: This leaves an unchecked conversion warning.

File: src/games/strategy/triplea/ui/mapdata/MapData.java [hunk 5]
```diff
@@ -193,43 +188,65 @@ public MapData(final ResourceLoader loader) {
...
-    final URL decorations = m_resourceLoader.getResource(DECORATIONS_FILE);
-    if (decorations == null) {
-      m_decorations = Collections.emptyMap();
-      return;
+  private Optional<Image> loadTerritoryNameImage(String imageName) {
+    Image img = null;
+    try {
+      // try first file names that have underscores instead of spaces
+      final String normalizedName = imageName.replace(' ', '_');
+      img = loadImage(constructTerritoryNameImagePath(normalizedName));
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/ui/mapdata/MapData.java]: Could we return Optional here to avoid the null?

File: src/games/strategy/triplea/ui/mapdata/MapData.java [hunk 6]
```diff
@@ -193,43 +188,65 @@ public MapData(final ResourceLoader loader) {
...
+      if (img == null) {
+        img = loadImage(constructTerritoryNameImagePath(imageName));
+      }
+    } catch(Exception e ) {
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/ui/mapdata/MapData.java]: Extra space after 'e'

File: src/games/strategy/triplea/ui/mapdata/MapData.java [hunk 7]
```diff
@@ -193,43 +188,65 @@ public MapData(final ResourceLoader loader) {
...
     }
-    m_decorations = new HashMap<>();
+    return Optional.ofNullable(img);
+  }
 
-    Optional<InputStream> inputStream = UrlStreams.openStream(decorations);
-    if(inputStream.isPresent()) {
+
+  private String constructTerritoryNameImagePath(String baseName) {
+    return "territoryNames/" + baseName+ ".png";
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/ui/mapdata/MapData.java]: Missing space after 'baseName'

File: src/games/strategy/triplea/ui/screen/IDrawable.java [hunk 8]
```diff
@@ -11,7 +11,7 @@
 import games.strategy.triplea.image.MapImage;
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/ui/screen/IDrawable.java]: Would the changes to this class have been better in a separate PR?

File: src/tools/map/making/MapProperties.java [hunk 9]
```diff
@@ -3,63 +3,25 @@
 import java.awt.Color;
```
Comment (ron-murhammer) [inline src/tools/map/making/MapProperties.java]: Would the changes to this class be better in a separate PR?

**Comment by ron-murhammer** [comment]
@DanVanAtta Comments added. I'd like to stress that having this broken up in a couple of PRs such as one for the territory name image fix, one for mapdata refactor, and one for each of any of the other class refactors would be much preferred.

File: src/games/strategy/triplea/ui/screen/IDrawable.java (hunk 8 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/screen/IDrawable.java]: The changes here are results of changes in MapData.
Note:
` final Optional<Point> namePlace = mapData.getNamePlacementPoint(territory);`
The return type is now optional. The changes here are strictly to deal with that.

File: src/games/strategy/triplea/ui/mapdata/MapData.java (hunk 2 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/mapdata/MapData.java]: The normalized one is the common case.

Below is a good approximate count of how many *png files we have with underscores and spaces:
```
dan@dan-desk:~/work/maps$ grep " " < <(find . -name "*png" -type f) | wc -l
1251
dan@dan-desk:~/work/maps$ grep "_" < <(find . -name "*png" -type f) | wc -l
99916
```
(note, counts are for 134 maps, that is most every map, but some very new ones might  not be included)

File: src/games/strategy/triplea/ui/mapdata/MapData.java (hunk 3 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/mapdata/MapData.java]: Agree, but I've been trying to bite off just one thing at a time per commit, rather than refactoring everything off the bat (as much fun as that would be). So simply just haven't gotten to that change yet - but there are a lot of things that need to change, and many of these variables need to be refactored away.

File: src/tools/map/making/MapProperties.java (hunk 9 above)
Comment (DanVanAtta) [inline src/tools/map/making/MapProperties.java]: No, they are dependent changes. Notice that in many of the removed methods we access a property of `MapData`. These often represent the only class-external usage of a `MapData`. Removing the unused methods then removes the only thing preventing those data members from being marked as private.

Removing everything unused does include some independent changes. It is included because removing unused code is a very beneficial cleanup and secondly keeping as one PR avoids the likely merge conflicts.

**Comment by DanVanAtta** [comment]
Basing multiple PRs on one another does not seem to be a very good practice in the end. Namely it causes merge conflicts, and does not allow for effective review of additional large changes until the base changes have been merged.

Instead, as was done here, one public branch served as the main branch with the bug fix and a simple file move. Given the fallacy of copy/paste, the initial bug fix was not correct, and I did not notice this until refactoring the 'fixed' method. Regardless, the incorrect bug fix turned a large problem into very small problem, and was better left as was. Given the lack of response to pointing out that the dependent branch had the full fix, I merged the two PRs together so the full fix would be in one PR. 

With that context, I am not now prepared to re-chop this up so we have a neat bug fix that conflicts with the rest of the changes. Chopping things up here I would estimate to take a few hours, and it is not clear to me along what different lines the commits can be structured, without ignoring large numbers of potential merge conflicts that would greatly depend on merge ordering.

**Comment by DanVanAtta** [comment]
To summarize, pushed some formatting changes, fixed the null stuff in the refactored `loadTerritoryNameImage` method, and fixed the unchecked conversion warning

**Comment by ron-murhammer** [comment]
@DanVanAtta Alright looks good. I see the dependencies on those other classes now. I know sometimes you already start refactoring then need to make a bug fix, etc. When possible definitely prefer to keep simple bug fixes in separate PRs so they don't end up having to wait on reviewing other changes and can be merged quickly.
